### ISSUE TITLE
SkiaImageManager LoadImageManagedAsync() && GetPendingItemLoadsForPriority() -silently- fail (random behavior)

### ISSUE THREAD

**Issue by sferhah** [taublast/DrawnUi.Maui#123]
**Screenshot:**
<img width="314" alt="skia_bug" src="https://github.com/user-attachments/assets/42f8132f-903f-4d67-948e-83ea795f6255">

It's very random, sometimes all nodes succeed in displaying the image, sometimes none of them (especially if there are many).

**Platforms:** WinUI & Android (iOs not tested).

Some -silent- stacktraces: 

```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.set_Item(TKey key, TValue value)
   at DrawnUi.Maui.Draw.SkiaImageManager.LoadImageManagedAsync(ImageSource source, CancellationTokenSource token, LoadPriority priority)
   at DrawnUi.Maui.Draw.SkiaImage.LoadImageManagedAsync(ImageSource source, CancellationTokenSource cancel, LoadPriority priority)
   at DrawnUi.Maui.Draw.SkiaImage.<>c__DisplayClass75_1.<<SetImageSource>g__LoadAction|1>d.MoveNext()
```

```
System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.set_Item(TKey key, TValue value)
   at DrawnUi.Maui.Draw.SkiaImageManager.LoadImageManagedAsync(ImageSource source, CancellationTokenSource token, LoadPriority priority)
   at DrawnUi.Maui.Draw.SkiaImage.LoadImageManagedAsync(ImageSource source, CancellationTokenSource cancel, LoadPriority priority)
   at DrawnUi.Maui.Draw.SkiaImage.<>c__DisplayClass75_1.<<SetImageSource>g__LoadAction|1>d.MoveNext()
```

```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Generic.Dictionary`2.Enumerator.MoveNext()
   at DrawnUi.Maui.Draw.SkiaImageManager.GetPendingItemLoadsForPriority(LoadPriority priority)
   at DrawnUi.Maui.Draw.SkiaImageManager.ProcessQueue()
```

**Workaround:**
Add an await Task.Delay() between each node addition.

**Comment by sferhah** [comment]
No hurry, I am just experimenting with your lib (so far the perfs are impressive !). Below you'll find the source code for a ContentPage you can add to the Sanbox github project:

```
using DrawnUi.Maui.Views;

namespace Sandbox.Views;

public partial class MainPageAnomalyNodes : ContentPage
{
    private SkiaLayout TreeLayout;
    public MainPageAnomalyNodes()
    {
        this.Content = new Canvas
        {
            Gestures = GesturesMode.Enabled,
            HorizontalOptions = LayoutOptions.Fill,
            VerticalOptions = LayoutOptions.Fill,
            Content = TreeLayout = new SkiaLayout
            {
                HorizontalOptions = LayoutOptions.Fill,
                VerticalOptions = LayoutOptions.Fill,
                BackgroundColor = Colors.LightGray,
            }
        };
    }

    bool firstAppear = true;
    protected override void OnAppearing()
    {
        base.OnAppearing();

        if (firstAppear)
        {
            firstAppear = false;

            foreach (NodeViewModel node in GenerateNodes())
            {
                AddNode(node);
            }
        }
    }

    public int GlobalOffset = 75;
    public const int NodeWidth = 40;
    public const int NodeHeight = 40;

    private void AddNode(NodeViewModel node)
    {
        Point point = node.Point.Offset(-(NodeWidth / 2), -(NodeHeight / 2)).Offset(GlobalOffset, GlobalOffset);

        TreeLayout.Children.Add(new SkiaButton
        {
            TranslationX = point.X,
            TranslationY = point.Y,
            Clicked = async (b, p) =>
            {
                await DisplayAlert("", node.Icon, "Ok");
            },
            Children =
            {
                new SkiaShape
                {
                    WidthRequest = NodeWidth,
                    HeightRequest = NodeHeight,
                    Type = ShapeType.Circle,
                    Background = node.Color,
                    BackgroundColor = node.Color,
                    Content = new SkiaImage
                    {
                        Aspect = TransformAspect.AspectFit,
                        Source = "Images/" + node.Icon,
                    }
                }
            }
        });
    }

    private NodeViewModel[] GenerateNodes()
    {
        return new NodeViewModel[]
        {
            new NodeViewModel
            {
                Label = "Root",
                Icon = "nico.jpg",
                Color = Colors.Blue,
                Point = new Point(0, 0),
            },
            new NodeViewModel
            {
                Label = "Child 1",
                Icon = "nico.jpg",
                Color = Colors.Red,
                Point = new Point(100,50)
            },
            new NodeViewModel
            {
                Label = "Child 1",
                Icon = "nico.jpg",
                Color = Colors.Red,
                Point = new Point(100,0)
            },
            new NodeViewModel
            {
                Label = "Child 1",
                Icon = "nico.jpg",
                Color = Colors.Red,
                Point = new Point(100,150)
            },
            new NodeViewModel
            {
                Label = "Child 2",
                Icon = "nico.jpg",
                Color = Colors.Green,
                Point = new Point(100,100)
            },
            new NodeViewModel
            {
                Label = "nico",
                Icon = "nico.jpg",
                Color = Colors.Red,
                Point = new Point(200,0)
            },
            new NodeViewModel
            {
                Label = "nico",
                Icon = "nico.jpg",
                Color = Colors.Red,
                Point = new Point(300,0)
            },
            new NodeViewModel
            {
                Label = "GrandChild 1",
                Icon = "nico.jpg",
                Color = Colors.Yellow,
                Point = new Point(150,100)
            },
            new NodeViewModel
            {
                Label = "GrandChild 2",
                Icon = "nico.jpg",
                Color = Colors.Yellow,
                Point = new Point(150,150)
            },
            new NodeViewModel
            {
                Label = "GrandChild 3",
                Icon = "nico.jpg",
                Color = Colors.Yellow,
                Point = new Point(150,210)
            },
            new NodeViewModel
            {
                Label = "Child 3",
                Icon = "nico.jpg",
                Color = Colors.Orange,
                Point = new Point(0,200)
            }
        };
    }
}


public class NodeViewModel
{
    required public string Label { get; set; }
    required public Color Color { get; set; }
    required public string Icon { get; set; }
    required public Point Point { get; set; }
}
```

**Comment by taublast** [comment]
Thanks so much!

**Comment by taublast** [comment]
So there was 2 issues inside the manager:
1 not all pending loads were unpaused
2 concurrent threads issue
Fixed with [c664efb3c67da36706dfb7c6853dd8c4469eb512](https://github.com/taublast/DrawnUi.Maui/commit/c664efb3c67da36706dfb7c6853dd8c4469eb512), now all images load fine, will be in next nuget, very soon.

Regarding your app use-case, there is a flag you might want to use `SkiaImageManager.ReuseBitmaps = true;`

> If set to true will not return clones for same sources, but will just return the existing cached SKBitmap reference. Useful if you have a lot on images reusing same sources, but you have to be carefull not to dispose the shared image. SkiaImage is aware of this setting and will keep a cached SKBitmap from being disposed.

You have repeating images so.. 

Another trick with images to use SkiaImage property `LoadSourceOnFirstDraw = false`. Normally for performance, we start loading image source only when the image is started first drawing. This override it for specific cases.

Also I saw you build your Ui in code behind, would invite you to check the repro page I was using, it was using HotReload for code-behind, could be useful. Isaw you are using OnAppearing fro MAUI, you have some stuff from DrawUI instead of this:
1 the canvas itself triggers  WillFirstTimeDraw (event)
2 any skiacontrol triggers LayoutIsReady when finally its height and width become >0, you can also override OnLayourReady for any control. I used this event in my repro:

https://github.com/taublast/DrawnUi.Maui/commit/77d099bc69ff5b41d980549ba8ea4a42a087512f
